src/System.IdentityModel.Tokens.Jwt/JwtSecurityTokenHandler.cs

```diff
@@ -600,20 +610,70 @@ public override SecurityToken CreateToken(SecurityTokenDescriptor tokenDescripto
             if (tokenDescriptor == null)
                 throw LogHelper.LogArgumentNullException(nameof(tokenDescriptor));
 
+            // TODO at next major version (8.0) use only Audiences as SecurityTokenDescriptor.Audience" will be removed.
+            string aud = tokenDescriptor.Audience;
+            IDictionary<string, object> claims = tokenDescriptor.Claims;
+
+            if (!tokenDescriptor.Audiences.IsNullOrEmpty())
+            {
+                claims = MergeClaimsAudValuesAndDescriptorAudiences(tokenDescriptor);
+                aud = null;
+            }
+
             return CreateJwtSecurityTokenPrivate(
                 tokenDescriptor.Issuer,
-                tokenDescriptor.Audience,
+                aud,
                 tokenDescriptor.Subject,
                 tokenDescriptor.NotBefore,
                 tokenDescriptor.Expires,
                 tokenDescriptor.IssuedAt,
                 tokenDescriptor.SigningCredentials,
                 tokenDescriptor.EncryptingCredentials,
-                tokenDescriptor.Claims,
+                claims,
                 tokenDescriptor.TokenType,
                 tokenDescriptor.AdditionalHeaderClaims,
                 tokenDescriptor.AdditionalInnerHeaderClaims);
         }
+        private static IDictionary<string, object> MergeClaimsAudValuesAndDescriptorAudiences(SecurityTokenDescriptor tokenDescriptor)
+        {
+            if (tokenDescriptor.Claims == null)
+                return new Dictionary<string, object>(){{JwtRegisteredClaimNames.Aud, tokenDescriptor.Audiences}};
+
+            if (tokenDescriptor.Claims.ContainsKey(JwtRegisteredClaimNames.Aud))
+                MergeAudClaimsIntoAudiencesProperty(tokenDescriptor);
+
+            var claimsCopy = new Dictionary<string, object>(tokenDescriptor.Claims);
```

Comment:
We need a solution without creating a new dictionary.

Reply:
For my own learning, what is the reason we need to avoid the new dictionary? Is it space constraints or something else?